firmware/rhizome.ino

```diff
@@ -98,22 +101,316 @@ void setup() {
+ * @param newPins The pins to check for
+ * @returns Whether any of the supplied pins are already in use
+ */
+bool arePinsInUse(std::list<int>* newPins) {
+    // Make sure ID isn't in use
+    std::deque<Ohmbrewer::Equipment*>::iterator itr = sprouts.begin();
+
+    // Iterate through all Sprouts
+    for (itr; itr != sprouts.end(); itr++) {
+            std::list<int> pinsInUse;
+            std::list<int>::iterator inusePinItr;
+            std::list<int>::iterator newPinItr;
+
+        // Iterate through all the pins for this Equipment...
+        (*itr)->whichPins(&pinsInUse);
+        for(inusePinItr = pinsInUse.begin(); inusePinItr != pinsInUse.end(); inusePinItr++) {
+            // Compare against all the pins for the new Equipment...
+            for(newPinItr = newPins->begin(); newPinItr != newPins->end(); newPinItr++) {
+                if((*newPinItr) == (*inusePinItr)) {
+                    return true;
+                }
+            }
+        }
+    }
+
+    return false;
+}
+
+/**
+ * Dynamically adds Equipment to the Rhizome.
+ * The Equipment will be added in the OFF state, with no current task.
+ * Hence, it is probably necessary to send an update immediately after a successful addition.
+ *
+ * Additionally, the argument string for this function must match the following format:
+ * TYPE,ID,{PINS}
+ * where
+ * TYPE matches the result of TYPE_NAME for the desired equipment class
+ * ID isn't already in use
+ * PINS is the appropriate length for the equipment type AND
+ *      is in the expected order for the equipment constructor that will be called (as described above)
+ *
+ * @param argsStr The argument string passed via the Particle Cloud.
+ * @returns Equipment ID if successful,
+ *          (negative) error codes if unsuccessful:
+ *          -1 : Invalid ID
+ *          -2 : ID already in use for given Equipment Type
+ *          -3 : One or more of the provided pins is already in use
+ *          -4 : Incorrect number of pins provided for given Equipment Type
+ */
+int addSprout(String argsStr) {
+    char* params = new char[argsStr.length() + 1];
+    strcpy(params, argsStr.c_str());
+
+    // Parse the parameters
+    String type = String(strtok(params, ","));
+    String idStr   = String(strtok(NULL, ","));
+    int id = idStr.toInt();
+
+    // If toInt() fails due to a bad parse, it gives 0.
+    if(isFakeZero(idStr)) {
+        delete params;
+        return -1;
+    }
+
+    // Make sure ID isn't in use
+    std::deque<Ohmbrewer::Equipment*>::iterator itr = sprouts.begin();
+
+    for (itr; itr != sprouts.end(); itr++) {
+        if (((*itr)->getID() == id) && (type.equalsIgnoreCase((*itr)->getType()))) {
+            delete params;
+            return -2;
+        }
+    }
+
+    // Now, depending on the type we need to parse differently. Then we add the Equipment.
+    if(type.equalsIgnoreCase(Ohmbrewer::Pump::TYPE_NAME)) {
+        String powerPin   = String(strtok(NULL, ","));
```

Comment:
this section will probably need to be gone over in more detail once the PID branch is merged since most/all pumps will currently be only power pin.